server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -589,6 +612,158 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
         }, listener::onFailure);
     }
 
+    /**
+     * Send requests to the target node to recover files from a given snapshot. In case of failure, the listener
+     * value contains the list of files that failed to be recovered from a snapshot.
+     */
+    void recoverSnapshotFiles(ShardRecoveryPlan shardRecoveryPlan,
+                              ActionListener<List<StoreFileMetadata>> listener) {
+        ShardRecoveryPlan.SnapshotFilesToRecover snapshotFilesToRecover = shardRecoveryPlan.getSnapshotFilesToRecover();
+
+        if (snapshotFilesToRecover.isEmpty()) {
+            listener.onResponse(Collections.emptyList());
+            return;
+        }
+
+        new SnapshotRecoverFileRequestsSender(shardRecoveryPlan, listener).start();
+    }
+
+    class SnapshotRecoverFileRequestsSender {
+        private final ShardRecoveryPlan.SnapshotFilesToRecover snapshotFilesToRecover;
+        final ActionListener<List<StoreFileMetadata>> listener;
+        final CountDown countDown;
+        final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> pendingSnapshotFilesToRecover;
+        final AtomicBoolean cancelled = new AtomicBoolean();
+        final Set<ListenableFuture<Void>> outstandingRequests = new HashSet<>(maxConcurrentSnapshotFileDownloads);
+        List<StoreFileMetadata> filesFailedToDownloadFromSnapshot;
```

Comment:
Looks like this can all be private?

```suggestion
    private class SnapshotRecoverFileRequestsSender {
        private final ShardRecoveryPlan.SnapshotFilesToRecover snapshotFilesToRecover;
        private final ActionListener<List<StoreFileMetadata>> listener;
        private final CountDown countDown;
        private final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> pendingSnapshotFilesToRecover;
        private final AtomicBoolean cancelled = new AtomicBoolean();
        private final Set<ListenableFuture<Void>> outstandingRequests = new HashSet<>(maxConcurrentSnapshotFileDownloads);
        private List<StoreFileMetadata> filesFailedToDownloadFromSnapshot;
```